lib/Sema/CSDiag.cpp

```diff
@@ -980,6 +981,69 @@ static bool argumentMismatchIsNearMiss(Type argType, Type paramType) {
   return false;
 }
 
+/// Given a parameter type that may contain generic type params and an actual
+/// argument type, decide whether the param and actual arg have the same shape
+/// and equal fixed type portions, and return by reference each archetype and
+/// the matching portion of the actual arg type where that archetype appears.
+static bool findGenericSubstitutions(Type paramType, Type actualArgType,
+                                     SmallVector<ArchetypeType *, 4> &archetypes,
+                                     SmallVector<Type, 4> &substitutions) {
+  if (auto genericParam = paramType->getAs<GenericTypeParamType>())
+    paramType = genericParam->getDecl()->getArchetype();
+  if (auto archetype = paramType->getAs<ArchetypeType>()) {
+    for (unsigned i = 0, c = archetypes.size(); i < c; i++) {
+      if (archetypes[i]->isEqual(archetype))
+        return substitutions[i]->isEqual(actualArgType);
+    }
+    archetypes.push_back(archetype);
+    substitutions.push_back(actualArgType);
+    return true;
+  }
+  
+  if (paramType->isEqual(actualArgType))
+    return true;
+  if (paramType->getKind() != actualArgType->getKind())
```

Comment:
Good direction for next commit. :)  I'm being conservative because failing to match just leads to the existing generic couldn't-call-with-these-args diagnosis, and I don't ever want to accidentally get too liberal, because then we'll emit a diagnosis which is more specific but possibly wrong.